homeassistant/components/assist_pipeline/pipeline.py

```diff
@@ -418,12 +421,29 @@ def process_event(self, event: PipelineEvent) -> None:
             return
         pipeline_data.pipeline_runs[self.pipeline.id][self.id].events.append(event)
 
-    def start(self) -> None:
+    async def start(self, device_id: str | None) -> None:
         """Emit run start event."""
         if self.debug_recording_dir is not None:
             # Create directory where wake/stt audio will be saved
-            _LOGGER.debug("Saving pipeline audio to %s", self.debug_recording_dir)
-            self.debug_recording_dir.mkdir(parents=True, exist_ok=True)
+            if device_id is None:
+                # <debug_recording_dir>/<pipeline.name>/<run.id>
+                self.debug_recording_dir_for_run = (
+                    self.debug_recording_dir / self.pipeline.name / self.id
+                )
+            else:
+                # <debug_recording_dir>/<device_id>/<pipeline.name>/<run.id>
+                self.debug_recording_dir_for_run = (
+                    self.debug_recording_dir / device_id / self.pipeline.name / self.id
+                )
+            _LOGGER.debug(
+                "Saving pipeline audio to %s", self.debug_recording_dir_for_run
+            )
+
+            def mkdir() -> None:
```

Comment:
We could delay this until we make the wav writer, saving a trip to the executor.

Reply:
Moved into the debug recording thread